### PR TITLE
Solutions for Chapter 4

### PR DESCRIPTION
**Pull request by keyridan**
### Solutions for Chapter 4
Wow, that was an amazing journey! Thank you! ❤️

Also, I have a small question about `andM` implementation, provided in `learn4haskell-solution-chapter4` branch. I was comparing my solution and found some strange behaviour that might be a bug. 
```
andM :: (Monad m) => m Bool -> m Bool -> m Bool
andM a b = a >>= \x -> if x then b else (pure False)
```
When I tried to use this version `andM` with Lists, I had a problem with losing elements, because `pure False` returns only one element and the second List could have more than one.
```
describe "Task 8*: Before the Final Boss andMWithListProblem" $ do
        it "True and False for small list" $ (Cons False Empty) `andMWithListProblem` (Cons False(Cons False Empty)) `shouldBe` (Cons False(Cons False Empty))
        --  expected: Cons False (Cons False Empty)
        --  but got: Cons False Empty
        it "True and False for list" $ (Cons True (Cons False Empty)) `andMWithListProblem` (Cons False(Cons True Empty)) `shouldBe` (Cons False(Cons True(Cons False (Cons False Empty))))
        -- expected: Cons False (Cons True (Cons False (Cons False Empty)))
        -- but got: Cons False (Cons True (Cons False Empty))
```
Is it by design or is it an actual bug?

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +486,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (Trap e) <*> _ = Trap e
+    (Reward f) <*> secret = fmap f secret
```
Comment (chshersh) [inline src/Chapter4.hs:493]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -628,8 +661,14 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+andMv1 :: (Monad m) => m Bool -> m Bool -> m Bool
+andMv1 a b = fmap (&&) a <*> b
```
Comment (chshersh) [inline src/Chapter4.hs:665]: If you don't use the `>>=` operator, then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.
```suggestion
andMv1 :: (Applicative m) => m Bool -> m Bool -> m Bool
andMv1 a b = fmap (&&) a <*> b
```

File: src/Chapter4.hs [hunk 3]
```diff
@@ -628,8 +661,14 @@ Can you implement a monad version of AND, polymorphic over any monad?
...
+
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = a >>= \x -> pure (x &&) <*> b
```
Comment (chshersh) [inline src/Chapter4.hs:668]: This version is similar to the previous one. Because `pure f <*> x` is the same as `fmap f x`

File: src/Chapter4.hs [hunk 4]
```diff
@@ -628,8 +661,14 @@ Can you implement a monad version of AND, polymorphic over any monad?
...
+
+andMWithListProblem :: (Monad m) => m Bool -> m Bool -> m Bool
+andMWithListProblem a b = a >>= \x -> if x then b else (pure False)
```
Comment (chshersh) [inline src/Chapter4.hs:671]: Answering your question regarding the behaviour. No, it's not a bug :slightly_smiling_face: A lazier version stops earlier. If you want to match all elements of lists pairwise, you need to use one of the previous implementations of `andM` that you've provided.

Basically, these are just different functions with different behaviours and for your real problem you should use the one that works for your best. However, I want to add that lists are rarely used as monads, even though they have a `Monad` instance. The monadic behaviour for lists has only very few use-cases, unlike data types like `Maybe` or `Either`. So most generic functions are usually optimized for these types, that's why you will find this lazy version of `andM` in most libraries out there.

File: test/Test/Chapter4.hs [hunk 5]
```diff
@@ -45,22 +45,78 @@ chapter4normal = describe "Chapter4Normal" $ do
...
         it "reward <*> reward" $
             Reward odd <*> Reward 42 `shouldBe` (Reward False :: Secret String Bool)
+    describe "Task5: Applicative for List" $ do
+        it "pure int" $ pure 1 `shouldBe` (Cons 1 Empty)
+        it "pure bool" $ pure True `shouldBe` (Cons True Empty)
+        it "pure false" $ pure False `shouldBe` (Cons False Empty)
+        it "Empty append list" $ Empty `append` list `shouldBe` list
+        it "list append empty" $ list `append` Empty `shouldBe` list
+        it "list append another list" $ list `append` anotherList `shouldBe` sumOfLists
+        it "list <*> list" $ fmap (+) list <*> anotherList `shouldBe` (Cons 3 (Cons 21 (Cons 12 (Cons 30 Empty))))
```
Comment (chshersh) [inline test/Test/Chapter4.hs:55]: Nice work on tests! :clap:

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 
I've answered your questions in my review, they should clarify some of the concepts better :slightly_smiling_face: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

File: src/Chapter4.hs (hunk 4 above)
Comment (keyridan) [reply src/Chapter4.hs:671]: Oh, I see! Thank you for the clarification! 😊

**Pull request merged**